src/turbomind/comm/thread_comm.cc

```diff
@@ -71,37 +68,50 @@ struct ThreadCommImpl: public HostCommImpl {
     std::shared_ptr<HostCommImpl> Split(int color, int key) override
     {
         TM_CHECK(color >= 0);
-        TM_CHECK(g2l_[rank_] >= 0);
-
-        // `g2l_[rank_]` imposes proper ordering when keys are equal
-        auto vec = comm::AllGather(this, std::make_tuple(color, key, g2l_[rank_]));
-
-        auto last = std::stable_partition(vec.begin(), vec.end(), [&](auto x) {  //
-            return std::get<0>(x) == color;
-        });
-        vec.erase(last, vec.end());
-        std::stable_sort(vec.begin(), vec.end(), [](auto& a, auto& b) {  //
-            return a < b;
-        });
-
-        std::vector<int> l2g;
-        std::vector<int> g2l(g2l_.size(), -1);
-
-        for (size_t i = 0; i < vec.size(); ++i) {
-            int r = l2g_.at(std::get<2>(vec[i]));
-            l2g.push_back(r);
-            g2l[r] = i;
+
+        auto ranks = comm::AllGather(this, std::make_tuple(color, key, rank_));
+
+        auto same_color = [&](auto x) { return std::get<0>(x) == color; };
+        ranks.erase(std::stable_partition(ranks.begin(), ranks.end(), same_color), ranks.end());
+
+        std::stable_sort(ranks.begin(), ranks.end(), [](auto& a, auto& b) { return a < b; });
+
+        std::shared_ptr<State> state;
+
+        int rank = -1;
+        for (int i = 0; i < ranks.size(); ++i) {
+            if (std::get<2>(ranks[i]) == rank_) {
+                rank = i;
+            }
+        }
+
+        TM_CHECK_GE(rank, 0);
+
+        if (rank == 0) {
+            state = std::make_shared<State>(ranks.size());
         }
 
-        return std::make_shared<ThreadCommImpl>(std::move(l2g), std::move(g2l), state_, rank_);
+        auto states = comm::AllGather(this, state);
+        if (rank != 0) {
+            const int root = std::get<2>(ranks[0]);
+            state          = states[root];
+        }
+
```

Comment:
Why not use comm::Broadcast ?

Reply:
The root is group specific. 
```
rank before split  0 1 2 3 4 5 6 7
     split color   0 0 0 0 1 1 1 1
rank after split   0 1 2 3 0 1 2 3
```
States are created by rank-0 after splitting. However boradcast operation has single root.